src/snowflake/snowpark/_internal/data_source/drivers/psycopg2_driver.py

```diff
@@ -0,0 +1,362 @@
+    Psycopg2TypeCode.UUID: StringType,
+    Psycopg2TypeCode.XML: StringType,
+    Psycopg2TypeCode.TSVECTOR: StringType,
+    Psycopg2TypeCode.TSQUERY: StringType,
+    Psycopg2TypeCode.TXID_SNAPSHOT: StringType,
+    Psycopg2TypeCode.PG_LSN: StringType,
+    Psycopg2TypeCode.PG_SNAPSHOT: StringType,
+}
+
+
+class Psycopg2Driver(BaseDriver):
+    def __init__(
+        self, create_connection: Callable[[], "Connection"], dbms_type: Enum
+    ) -> None:
+        super().__init__(create_connection, dbms_type)
+
+    def to_snow_type(self, schema: List[Any]) -> StructType:
+        # The psycopg2 spec is defined in the following links:
+        # https://www.psycopg.org/docs/cursor.html#cursor.description
+        # https://www.psycopg.org/docs/extensions.html#psycopg2.extensions.Column
+        fields = []
+        for (
+            name,
+            type_code,
+            _display_size,
+            _internal_size,
+            precision,
+            scale,
+            _null_ok,
+        ) in schema:
+            try:
+                type_code = Psycopg2TypeCode(type_code)
+            except ValueError:
+                raise NotImplementedError(
+                    f"Postgres type not supported: {type_code} for column: {name}"
+                )
+            snow_type = BASE_POSTGRES_TYPE_TO_SNOW_TYPE.get(type_code)
+            if snow_type is None:
+                raise NotImplementedError(
+                    f"Postgres type not supported: {type_code} for column: {name}"
+                )
+            if Psycopg2TypeCode(type_code) == Psycopg2TypeCode.NUMERICOID:
+                if not self.validate_numeric_precision_scale(precision, scale):
+                    logger.debug(
+                        f"Snowpark does not support column"
+                        f" {name} of type {type_code} with precision {precision} and scale {scale}. "
+                        "The default Numeric precision and scale will be used."
+                    )
+                    precision, scale = None, None
+                data_type = snow_type(
+                    precision if precision is not None else 38,
+                    scale if scale is not None else 0,
+                )
+            elif type_code == Psycopg2TypeCode.TIMESTAMPTZOID:
+                data_type = snow_type(TimestampTimeZone.TZ)
+            else:
+                data_type = snow_type()
+            fields.append(StructField(name, data_type, True))
+        return StructType(fields)
+
+    @staticmethod
+    def data_source_data_to_pandas_df(
+        data: List[Any], schema: StructType
+    ) -> "pd.DataFrame":
+        df = BaseDriver.data_source_data_to_pandas_df(data, schema)
+
+        variant_type_indexes = [
+            i
+            for i, field in enumerate(schema.fields)
+            if isinstance(field.datatype, VariantType)
+        ]
+        col_names = df.columns[variant_type_indexes]
+        df[col_names] = BaseDriver.df_map_method(df[col_names])(
```

Comment:
just curious, does postgres has functionality similar to outputtypehandler in oracledb that could allow us to do the conversion in it?

Reply:
I updated the code to use pushdown for all the types